test/Parse/diagnostic_missing_func_keyword.swift

```diff
@@ -27,9 +27,9 @@ infix operator %%
 struct Bar {
   fisr = 0x5F3759DF // expected-error {{expected 'var' keyword in property declaration}} {{3-3=var }}
 
-  %%<T: Brew> (lhs: T, rhs: T) -> T { // expected-error {{expected 'func' keyword in operator function declaration}} {{3-3=func }}
-                                      // expected-error @-1 {{operator '%%' declared in type 'Bar' must be 'static'}}
-                                      // expected-error @-2 {{member operator '%%' must have at least one argument of type 'Bar'}}
+  %%<T: Brew> (lhs: T, rhs: T) -> T { // expected-error {{expected declaration}}
```

Comment:
I don't understand why this regressed... could you describe why that's the case? I think this case and the one below:

```
a, b: Int // expected-error {{expected 'var' keyword in property declaration}} {{3-3=var }}
```

should still continue to work as they do right now.

Reply:
yes, I'm concerned about that right now. I have to retrace all of this to try to figure it out.

Reply:
Do you have the Xcode debugger working for you? (IIRC you are using Xcode?). That should help in figuring out the flow. I'm not sure if we have the steps written down somewhere (this talk https://youtu.be/oGJKsp-pZPk describes the setup, except that now we use the `swift-frontend` scheme instead of the `swift` scheme).

Reply:
Hypothesis is that when the possibly-func logic handles `%%<` it does not cleanly use all the tokens up to the following right brace and the following parses are confused.

Reply:
I just use `lldb` directly since I build with ninja.

If I run the test with the built compiler in lldb and breakpoint where `IsProbablyVarDecl` is declared at the top of the `if (Flags.contains(PD_HasContainerType) && IsAtStartOfLineOrPreviousHadSemi) {` code block, and print `Tok` and then process tokens until I get to the `%%`,  I can see that the next token that we reach in this part of code after the `%%` is the `abc` in `abc : Int = 10`. The tokens in between are processed somewhere else, probably in `parseFunc` although not 100% sure on that. It looks like it's just swallowing up until the last right brace instead of the one at the end of the operator definition. I might still  be misinterpreting the behavior--I have not analyzed it that deeply yet. 

I'm not sure why the change knocks it out of balance though. 

restatement of rules: 

if we have something that looks like a `var` declaration without the `var` keyword, emit a diagnostic then try to parse the rest of it as a `var` without the keyword.

if we have something that looks like a `func` without the keyword, emit a diagnostic then try to parse the rest of it as a func without the keyword. 

otherwise just emit the 'expected declaration' diagnostic. 

The original behavioral problem is that `identifier` `period` is being treated as a possible func, then when it tries to parse the rest of it as a function it gets really confused and you get a diagnostic cascade. 

another possibility here is to catch the `identifier` `period` (`start` `identifier` `period` I guess) only, then generate the simple diagnostic for it and break out. Keep the original possible-var and possible-func logic for all other cases. Not ideal but should clear this up. I am not sure I have actually done that so far while wrestling with this.

Reply:
I reverted this back to a simple approach that emits a diagnostic and breaks when we see the id-period pattern. But In this version it does not quite work--if you put two of them in a row as in the test, the second one does not raise the diagnostic, probably because I need to parse the rest of the id-period faux-expression to be ready for the following one. 

I think I need to parse from the first token as far as the parser goes for an id.id pattern and discard the result, more or less the way the others do. Let me try to figure out how to do that. There might be some context management I have to do to make it work.

Reply:
I really think the `IsProbablyFuncDecl` approach is the right track to be on here. It's going to cover *way* more categories of mistakes than identifier-period would.

If you look in `Parser::parseDeclFunc()`, you'll see that it splits trailing `<` characters out of operator tokens:

https://github.com/apple/swift/blob/master/lib/Parse/ParseDecl.cpp#L6307-L6319

You can either incorporate similar logic into the heuristic (allowing either identifier/operator tokens followed by a `tok::l_paren` or `tok::l_angle`, or two-or-more-character operator tokens with a trailing `<` followed by an identifier), or you can accept the regression in this test.

Reply:
Thanks! If we go with `IsProbablyFuncDecl` you mean that we'd make that test more specific, right? Then the identifier-dot problem would just fall out the bottom to the other expected-declaration diagnostic. The `parseDeclFunc` code would be the model to follow for evaluating `IsProbablyFuncDecl`. Did I understand that right? 

We might still have the problem that my test has (because the first identifier-period will go all the way to the right brace) but that does not bother me as much as making the test for identifier-period over specific when there's a better way to fix it.

Reply:
> If we go with IsProbablyFuncDecl you mean that we'd make that test more specific, right? Then the identifier-dot problem would just fall out the bottom to the other expected-declaration diagnostic.

Yes, exactly.

It looks like, in most cases, `parseDecl()`'s caller will skip to the next statement if `parseDecl()` returns an error, so you shouldn't need any special recovery in `parseDecl()` itself—just declare bankruptcy, return `DeclResult` without setting it to anything (it's an error by default), and let the caller worry about cleaning up the mess.